### DISCUSSION TITLE
Current State of Download Stats / Usage Statistics

### DISCUSSION THREAD

**Discussion by joergrech** [verdaccio/verdaccio#3213] [Ideas]
Hi, I want to collect same info on the current state of **Download Stats** / **Usage Statistics** (a.k.a., "Weekly Downloads" as shown in npmjs.com) mentioned in https://github.com/verdaccio/verdaccio/issues/1606, https://github.com/verdaccio/verdaccio/issues/1201, and https://github.com/verdaccio/verdaccio/issues/1482. 

As far as i see no plan exists to implement it in version 5.x or 6.x and demand is low (Only @fkyq, @favoyang, @chentianyuan, @sharonlx, @porcelainHeart, and I seem to be interested).

However, maybe we can use this "Discussion" to collect some requirements and starting points for such a feature so that someone (maybe me) can implement at least a rudimentary version.

## Metrics

From my point of view, the following metrics are interesting:
* M1: Count `Number of downloads per package`
* M2: Count `Number of downloads per package per user`

## General Requirements

* A place to count the event(s)
  * Option 1: Somewhere in verdaccio itself
  * Option 2: Counting in the `allow_access()` function of an Auth plugin
    * Contra: not precise - will be called not only for a download (`npm install`) activities
* A place to store the data
  * Option 1: A new database (esp. for time-series / histograms or more dynamic data analytics)
    * Contra: does not exist yet (requires an additional system)
    * Pro: allows time-series
  * Option 2: Store raw access data in files within the storage directory (e.g., using a file such as `access.csv`)
    * Contra: Might not support user-centric metrics (M2)
  * Option 3: Within the "User" database of an auth plugin
    * Contra: can be ephemeral (e.g., when using `verdaccio-htpasswd`)
    * Pro: already exists; can be persistent (e.g., when using `verdaccio-mongodb`)
* A time and place to aggregate the data for a time period (e.g., weekly)
  * Option 1: Init aggregate when website is called (but only once per period)
  * Option 2: External cronjob aggregating the data
* A way to access the data in the web UI
  * Option 1: Integrate access to database
  * Option 2: Store (aggregated) data with manifest in storage (e.g., using a file such as `metrics.json`)

### PS1: Using an Auth-Plugin with an attached database
* Example: Counting in the `verdaccio-mongodb` auth plugin
  * I've integrated a count into my MongoDB Auth plugin (1.1.6+) and observed the following:
    * Counting all access, publish, and unpublish activities of packages (M1) and for users (M2) is possible (but not 100% precise)
    * Counting from `allow_access` will count `npm install`, `npm bugs`, `npm deprecate`, `npm docs`, `npm show`, etc. (but not publish and unpublish).
    * Counting from `allow_publish` will count only `npm publish` (but also failed publish attempts - e.g., when version was already previously published or package has wrong format).
    * Counting from `allow_unpublish` will count only `npm unpublish` (but also failed unpublish attempts - e.g., when version was already previously unpublished).
    * Access or visualization of the data is only possible in the MongoDB (or an attached web app) - not in verdaccio itself.
  * Pro: 
    * No need for additional database
  * Contra: 
    * Precise counting of downloads (`npm install`) requires new parameters for `allow_access()` specifying the npm command such as `install`, `bugs`, `show`, etc. (e.g., `allow_access(user: RemoteUser, pkg: PackageAccess, cb: AuthAccessCallback, cmd: string)`). This would also allow for a finer-grained security definition (i.e., someone might be allowed to `npm install` but not to `npm deprecate`
    * Visualization in the web interface requires access to the database via the auth plugin (e.g., by introducing a `get_package_stats(pkg: PackageAccess)` or `get_user_stats(user: RemoteUser)` function in the auth plugin)

Did I miss something obvious? Can a middleware-plugin be used similar to `verdaccio-openmetrics`? Do you have other ideas to implement such a feature?

**Comment by edclement** [comment]
Hi Joe, In my opinion, metrics should be tracked using a standard/proven tool. If you're familiar with [Prometheus](https://prometheus.io/), this plugin might be helpful for what you are looking to accomplish: https://github.com/xlts-dev/verdaccio-prometheus-middleware
